port/unix/omrsysinfo.c

```diff
@@ -2004,6 +2006,68 @@ omrsysinfo_get_OS_version(struct OMRPortLibrary *portLibrary)
 					rc = sysctlbyname(KERN_OSPRODUCTVERSION, buffer, &resultSize, NULL, 0);
 				}
 				buffer[resultSize] = '\0';
+
+				/* 10.16 is the version returned when in a backwards compatibility context in OSX versions 11+
+				* so we check the plink symlink as that bypasses the compatibility context and returns the
+				* real version info.
+				*/
+				if (0 == strncmp(buffer, "10.16", 5)) {
+					portLibrary->mem_free_memory(portLibrary, buffer);
+					buffer = NULL;
+					len = 0;
+					portLibrary->mem_free_memory(portLibrary, allocatedFileBuffer);
+					allocatedFileBuffer = NULL;
+
+					intptr_t plistFile = portLibrary->file_open(portLibrary, COMPAT_PLIST_LINK, EsOpenRead, 0444);
+					if (plistFile >= 0) {
+						char myBuffer[1024];  /* the file size is ~500 bytes, 525 on a 11.3 */
+						char *fileBuffer = myBuffer;
+						len = (size_t) portLibrary->file_flength(portLibrary, plistFile);
+						rc = 0;
+						if (len >= sizeof(myBuffer)) { /* leave space for a terminating null */
+							allocatedFileBuffer = portLibrary->mem_allocate_memory(portLibrary, len + 1,
+								OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
+							fileBuffer = allocatedFileBuffer;
+						}
+						if (NULL != fileBuffer) {
```

Comment:
There is a bug here (and the previous code). If file_length() fails then `fileBuffer != NULL` and len is negative and the `fileBuffer[len] = '\0'` below writes outside the bounds of the stack buffer. Also strstr() will be checking random data and may go outside the bounds of readable memory, or find a false hit (however unlikely) causing another write outside of the buffer.

Reply:
Updated it so we look at the return value of `file_flength()` first before assigning `fileBuffer`. If the return is an error value, `fileBuffer` remains NULL.

Reply:
Looks good, but how about assigning directly the fileLen, I don't why rc needs to be involved.

Reply:
The return value of `file_flength()` is negative when an error occurs, and `fileLen` is an unsigned integer type.

Reply:
size_t doesn't need to be used, file_flength() returns int64_t and the code could just keep using that. I'll defer to @keithc-ca